### DISCUSSION TITLE
How can you create a plugin with subcommands?

### DISCUSSION THREAD

**Discussion by mtkennerly** [python-poetry/poetry#8091] [Q&A] [answered]
I haven't quite managed to figure this out from the Poetry or Cleo docs. I'd like to update [my plugin](https://github.com/mtkennerly/poetry-dynamic-versioning/blob/master/poetry_dynamic_versioning/plugin.py) to allow usage like this:

```
poetry dynamic-versioning apply --foo
poetry dynamic-versioning enable --bar
```

Is this supported? Any help would be appreciated.

**Comment by edgarrmondragon** [comment]
[This issue](https://github.com/python-poetry/cleo/issues/62) led me to [this example](https://github.com/pmav99/cleo_subcommands_example/blob/master/myapp/console/commands/cmd1/__init__.py).

It's a bit old but hopefully the API hasn't changed too much.

**Comment by mtkennerly** [comment, marked as answer]
I wasn't able to make that example work, unfortunately, but thank you for the suggestion.

I did have the idea to check how Poetry's own subcommands work, and I noticed this:

https://github.com/python-poetry/poetry/blob/6e942983dff1bcc6d307c7704e8159df0c959a16/src/poetry/console/commands/debug/info.py#L8-L9

So for a Poetry plugin, that looks like this:

```python
class DynamicVersioningEnableCommand(Command):
    name = "dynamic-versioning enable"

# ...

        application.command_loader.register_factory(
            "dynamic-versioning enable", lambda: DynamicVersioningEnableCommand(application)
        )
```

And it works!

```
$ poetry list
[...]
Available commands:
[...]
  config                     Manages configuration settings.
  dynamic-versioning         Apply the dynamic version to all relevant files and leave the changes in-place. This allows you to activate the plugin behavior on demand and inspect the result.
  export                     Exports the lock file to alternative formats.
[...]

 dynamic-versioning
  dynamic-versioning enable  Update pyproject.toml to enable dynamic versioning.

[...]
```

The only downside (in my opinion) is that `poetry dynamic-versioning --help` doesn't mention the subcommand.
